File: src/storage/r2Adapter.js
/**
 * Cloudflare R2 storage adapter for Node.js using AWS SDK v3
 */

import { S3Client, GetObjectCommand, PutObjectCommand, HeadObjectCommand, ListObjectsV2Command, DeleteObjectCommand } from '@aws-sdk/client-s3';
import { StorageAdapter } from './adapter.js';

export class R2StorageAdapter extends StorageAdapter {
    constructor(config) {
        super();
        this.bucketName = config.bucketName;
        this.environment = config.environment || 'prod';

        // Initialize S3 client with R2 endpoint
        this.s3Client = new S3Client({
            region: config.region || 'auto',
            endpoint: config.endpointUrl,
            credentials: {
                accessKeyId: config.accessKeyId,
                secretAccessKey: config.secretAccessKey,
            },
        });
    }

    _addEnvironmentPrefix(objectKey) {
        // Remove leading slashes
        const cleanKey = objectKey.replace(/^[/\\]+/, '');
        return `${this.environment}/${cleanKey}`;
    }

    _removeEnvironmentPrefix(fullKey) {
        const prefix = `${this.environment}/`;
        if (fullKey.startsWith(prefix)) {
            return fullKey.substring(prefix.length);
        }
        return fullKey;
    }

    async readFile(objectKey) {
        const fullKey = this._addEnvironmentPrefix(objectKey);
        try {
            const command = new GetObjectCommand({
                Bucket: this.bucketName,
                Key: fullKey,
            });
            const response = await this.s3Client.send(command);

            // Convert stream to buffer
            const chunks = [];
            for await (const chunk of response.Body) {
                chunks.push(chunk);
            }
            return Buffer.concat(chunks);
        } catch (error) {
            if (error.name === 'NoSuchKey') {
                throw new Error(`File not found in R2: ${objectKey}`);
            }
            throw new Error(`Failed to read from R2: ${objectKey} - ${error.message}`);
        }
    }

    async writeFile(objectKey, content) {
        const fullKey = this._addEnvironmentPrefix(objectKey);
        console.log(`[R2Adapter] Writing file to R2`);
        console.log(`[R2Adapter]   Object key: ${objectKey}`);
        console.log(`[R2Adapter]   Full key (with env): ${fullKey}`);
        console.log(`[R2Adapter]   Bucket: ${this.bucketName}`);
        console.log(`[R2Adapter]   Environment: ${this.environment}`);
        console.log(`[R2Adapter]   Size: ${Math.round(content.length / 1024)}KB`);

        try {
            const command = new PutObjectCommand({
                Bucket: this.bucketName,
                Key: fullKey,
                Body: content,
            });
            await this.s3Client.send(command);
            console.log(`[R2Adapter] ✅ File uploaded successfully to R2: ${fullKey}`);
        } catch (error) {
            console.error(`[R2Adapter] ❌ Failed to write to R2: ${objectKey}`);
            console.error(`[R2Adapter] Error details:`, error.message);
            console.error(`[R2Adapter] Error code:`, error.name);
            throw new Error(`Failed to write to R2: ${objectKey} - ${error.message}`);
        }
    }

    async fileExists(objectKey) {
        const fullKey = this._addEnvironmentPrefix(objectKey);
        try {
            const command = new HeadObjectCommand({
                Bucket: this.bucketName,
                Key: fullKey,
            });
            await this.s3Client.send(command);
            return true;
        } catch (error) {
            if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
                return false;
            }
            return false;
        }
    }

    async getFileSize(objectKey) {
        const fullKey = this._addEnvironmentPrefix(objectKey);
        try {
            const command = new HeadObjectCommand({
                Bucket: this.bucketName,
                Key: fullKey,
            });
            const response = await this.s3Client.send(command);
            return response.ContentLength;
        } catch (error) {
            if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
                throw new Error(`File not found in R2: ${objectKey}`);
            }
            throw new Error(`Failed to get file size from R2: ${objectKey} - ${error.message}`);
        }
    }

    async listFiles(prefix) {
        const fullPrefix = this._addEnvironmentPrefix(prefix);
        const files = [];

        try {
            let continuationToken = undefined;

            do {
                const command = new ListObjectsV2Command({
                    Bucket: this.bucketName,
                    Prefix: fullPrefix,
                    ContinuationToken: continuationToken,
                });

                const response = await this.s3Client.send(command);

                if (response.Contents) {
                    for (const obj of response.Contents) {
                        // Remove environment prefix from returned keys
                        const key = this._removeEnvironmentPrefix(obj.Key);
                        files.push(key);
                    }
                }

                continuationToken = response.NextContinuationToken;
            } while (continuationToken);

        } catch (error) {
            throw new Error(`Failed to list files in R2: ${prefix} - ${error.message}`);
        }

        return files;
    }

    async deleteFile(objectKey) {
        const fullKey = this._addEnvironmentPrefix(objectKey);
        try {
            const command = new DeleteObjectCommand({
                Bucket: this.bucketName,
                Key: fullKey,
            });
            await this.s3Client.send(command);
        } catch (error) {
            throw new Error(`Failed to delete file from R2: ${objectKey} - ${error.message}`);
        }
    }
}

File: workers/storage/r2_adapter.py
"""Cloudflare R2 storage adapter using S3-compatible API."""

from __future__ import annotations

from typing import List

from workers.core.errors import PermanentError, RetryableError
from workers.storage.adapter import StorageAdapter

# boto3 is optional - only required when using R2
try:
    import boto3
    from botocore.exceptions import ClientError, NoCredentialsError
    _BOTO3_AVAILABLE = True
except ImportError:
    _BOTO3_AVAILABLE = False


class R2StorageAdapter(StorageAdapter):
    """Storage adapter for Cloudflare R2 (S3-compatible)."""

    def __init__(
        self,
        account_id: str,
        access_key_id: str,
        secret_access_key: str,
        bucket_name: str,
        environment: str = "prod",
        region: str = "auto",
    ):
        """
        Initialize R2 storage adapter.
        
        Args:
            account_id: Cloudflare account ID
            access_key_id: R2 access key ID
            secret_access_key: R2 secret access key
            bucket_name: R2 bucket name
            environment: Environment prefix (e.g., "prod", "dev", "staging")
            region: R2 region (default: "auto")
        """
        if not _BOTO3_AVAILABLE:
            raise ImportError(
                "boto3 is required for R2 storage. Install with: pip install boto3"
            )

        self.bucket_name = bucket_name
        self.environment = environment
        
        # Construct R2 endpoint URL
        endpoint_url = f"https://{account_id}.r2.cloudflarestorage.com"
        
        # Initialize S3 client with R2 endpoint
        self.s3_client = boto3.client(
            "s3",
            endpoint_url=endpoint_url,
            aws_access_key_id=access_key_id,
            aws_secret_access_key=secret_access_key,
            region_name=region,
        )

    def _add_environment_prefix(self, object_key: str) -> str:
        """Add environment prefix to object key."""
        # Remove leading slashes
        clean_key = object_key.lstrip("/")
        return f"{self.environment}/{clean_key}"

    def _remove_environment_prefix(self, full_key: str) -> str:
        """Remove environment prefix from object key."""
        prefix = f"{self.environment}/"
        if full_key.startswith(prefix):
            return full_key[len(prefix):]
        return full_key

    def read_file(self, object_key: str) -> bytes:
        """Read file from R2."""
        full_key = self._add_environment_prefix(object_key)
        try:
            response = self.s3_client.get_object(Bucket=self.bucket_name, Key=full_key)
            return response["Body"].read()
        except ClientError as exc:
            error_code = exc.response.get("Error", {}).get("Code", "")
            if error_code == "NoSuchKey":
                raise PermanentError(f"File not found in R2: {object_key}") from exc
            elif error_code in ("AccessDenied", "InvalidAccessKeyId"):
                raise PermanentError(f"Permission denied for R2: {object_key}") from exc
            else:
                raise RetryableError(f"Failed to read from R2: {object_key}") from exc
        except NoCredentialsError as exc:
            raise PermanentError("R2 credentials not configured") from exc
        except Exception as exc:
            raise RetryableError(f"Failed to read from R2: {object_key}") from exc

    def write_file(self, object_key: str, content: bytes) -> None:
        """Write file to R2."""
        full_key = self._add_environment_prefix(object_key)
        try:
            self.s3_client.put_object(
                Bucket=self.bucket_name,
                Key=full_key,
                Body=content,
            )
        except ClientError as exc:
            error_code = exc.response.get("Error", {}).get("Code", "")
            if error_code in ("AccessDenied", "InvalidAccessKeyId"):
                raise PermanentError(f"Permission denied for R2: {object_key}") from exc
            else:
                raise RetryableError(f"Failed to write to R2: {object_key}") from exc
        except NoCredentialsError as exc:
            raise PermanentError("R2 credentials not configured") from exc
        except Exception as exc:
            raise RetryableError(f"Failed to write to R2: {object_key}") from exc

    def file_exists(self, object_key: str) -> bool:
        """Check if file exists in R2."""
        full_key = self._add_environment_prefix(object_key)
        try:
            self.s3_client.head_object(Bucket=self.bucket_name, Key=full_key)
            return True
        except ClientError as exc:
            error_code = exc.response.get("Error", {}).get("Code", "")
            if error_code in ("404", "NoSuchKey"):
                return False
            # Other errors are treated as "unknown"
            return False
        except Exception:
            return False

    def get_file_size(self, object_key: str) -> int:
        """Get file size from R2."""
        full_key = self._add_environment_prefix(object_key)
        try:
            response = self.s3_client.head_object(Bucket=self.bucket_name, Key=full_key)
            return response["ContentLength"]
        except ClientError as exc:
            error_code = exc.response.get("Error", {}).get("Code", "")
            if error_code in ("404", "NoSuchKey"):
                raise PermanentError(f"File not found in R2: {object_key}") from exc
            else:
                raise RetryableError(f"Failed to get file size from R2: {object_key}") from exc
        except Exception as exc:
            raise RetryableError(f"Failed to get file size from R2: {object_key}") from exc

    def list_files(self, prefix: str) -> List[str]:
        """List files with given prefix in R2."""
        full_prefix = self._add_environment_prefix(prefix)
        try:
            files = []
            paginator = self.s3_client.get_paginator("list_objects_v2")
            for page in paginator.paginate(Bucket=self.bucket_name, Prefix=full_prefix):
                for obj in page.get("Contents", []):
                    # Remove environment prefix from returned keys
                    key = self._remove_environment_prefix(obj["Key"])
                    files.append(key)
            return files
        except Exception as exc:
            raise RetryableError(f"Failed to list files in R2: {prefix}") from exc

    def delete_file(self, object_key: str) -> None:
        """Delete file from R2."""
        full_key = self._add_environment_prefix(object_key)
        try:
            self.s3_client.delete_object(Bucket=self.bucket_name, Key=full_key)
        except Exception as exc:
            raise RetryableError(f"Failed to delete file from R2: {object_key}") from exc
